Fix panic for provides a set to the key of a precondition and deny condition

vyankyGH opened the pull request:
Signed-off-by: Vyankatesh vyankateshkd@gmail.com
closes : https://github.com/kyverno/kyverno/issues/3147

## Milestone of this PR
/milestone 1.6.0

## What type of PR is this
 /kind bug
<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->



<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->



### Proof Manifests
```
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: drop-cap-net-raw
spec:
  validationFailureAction: enforce
  background: false
  rules:
  - name: drop-cap-net-raw
    match:
      resources:
        kinds:
        - Pod
    validate:
      deny:
        conditions:
          any:
          - key: 
            - "30"
            - "40"
            operator: AnyIn
            value: 10-100
```
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - name: test
    image: defdasdabian:923
```
Result:
```
Applying 1 policy to 1 resource... 
(Total number of result count may vary as the policy is mutated by Kyverno. To check the mutated policy please try with log level 5)

policy drop-cap-net-raw -> resource default/Pod/test failed: 
1. drop-cap-net-raw: validation error: rule drop-cap-net-raw failed 

pass: 0, fail: 1, warn: 0, error: 0, skip: 2 
```
## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the doc update and the link is:
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

realshuting commented on pkg/policy/validate.go:
```diff
@@ -1004,6 +1004,16 @@ func validateConditionValues(c kyverno.Condition) (string, error) {
 		return "", fmt.Errorf("entered value of `key`, `value` or `operator` is missing or misspelled")
 	}
+	switch reflect.TypeOf(c.Key).Kind() {
+	case reflect.String:
```
What about arrays? We need to validate list items as well.

vyankyGH replied:
This function is to validating  value for "{{request.operation}} " variable . It not make sense to add {{request.operation}} is in array as value of variable will be single value . So not added for array.

samj1912 commented:
@vyankyGH should we also add the proof manifest as a CLI test?

vyankyGH commented:
> @vyankyGH should we also add the proof manifest as a test(CLI or e2e)?

I have already added proof manifest for CLI apply command. Should  add for Test command?

samj1912 commented:
> > @vyankyGH should we also add the proof manifest as a test(CLI or e2e)?
> 
> I have already added proof manifest for CLI apply command. Should add for Test command? Asking in PR or in policies?

In the PR under something like https://github.com/kyverno/kyverno/tree/main/test/cli/test - helps make it easier to prevent regressions :)

chipzoller commented:
> > > @vyankyGH should we also add the proof manifest as a test(CLI or e2e)?
> > 
> > 
> > I have already added proof manifest for CLI apply command. Should add for Test command? Asking in PR or in policies?
> 
> In the PR under something like https://github.com/kyverno/kyverno/tree/main/test/cli/test - helps make it easier to prevent regressions :)

Absolutely, and we need many tests here, including the wildcard tests.

samj1912 commented on pkg/policy/validate.go:
```diff
@@ -1004,6 +1004,16 @@ func validateConditionValues(c kyverno.Condition) (string, error) {
 		return "", fmt.Errorf("entered value of `key`, `value` or `operator` is missing or misspelled")
 	}
+	switch reflect.TypeOf(c.Key).Kind() {
+	case reflect.String:
+		value, err := validateValuesKeyRequest(c)
```
```suggestion
	case reflect.String:
		// we currently only validate the key if it is a string (to validate the `request.operation`)
		// and for other kind of keys, we don't have any pre-defined validations
		value, err := validateValuesKeyRequest(c)
```

vyankyGH commented:
> > > @vyankyGH should we also add the proof manifest as a test(CLI or e2e)?
> > 
> > 
> > I have already added proof manifest for CLI apply command. Should add for Test command? Asking in PR or in policies?
> 
> In the PR under something like https://github.com/kyverno/kyverno/tree/main/test/cli/test - helps make it easier to prevent regressions :)

added for both Pass and fail testcase.

vyankyGH commented:
> > > > @vyankyGH should we also add the proof manifest as a test(CLI or e2e)?
> > > 
> > > 
> > > I have already added proof manifest for CLI apply command. Should add for Test command? Asking in PR or in policies?
> > 
> > 
> > In the PR under something like https://github.com/kyverno/kyverno/tree/main/test/cli/test - helps make it easier to prevent regressions :)
> 
> Absolutely, and we need many tests here, including the wildcard tests. [This ](https://github.com/kyverno/policies/issues/203)can also be addressed there.

creating Separate PR.

samj1912 commented:
@vyankyGH will you also be back porting this to 1.6 branch?

realshuting commented:
@samj1912 - can we add this to 1.6.0? I'll cherry-pick this fix into release-1.6.
